Roles improvements

Stankman opened the pull request.

Commit:
More RankRole settings

Commit:
Ares Backend Big Update

Commit:
RoleHierarchy Changes

Commit:
Assign Rank to Roles Done

shubrt commented on .vscode/launch.json:
```diff
@@ -0,0 +1,48 @@
+{
```
VS Code configuration files should be excluded by .gitignore

shubrt commented on src/Permission/Controller/PermissionController.php:
```diff
@@ -36,14 +43,98 @@ public function __construct(
+     * @return Response
+     * @throws DataObjectManagerException
+     */
+    public function rankById(Request $request, Response $response, array $args): Response
```
Please add a short description to each function doc block.

Stankman replied:
Many methods on _RoleController_ and _PermissionController_ require renaming + separate RoleHierarchy Functions, RoleRank Functions just as _RolePermissions_

shubrt commented on src/Permission/Repository/PermissionRepository.php:
```diff
@@ -41,4 +41,33 @@ public function getListOfUserWithRanks(): Collection
+    }
+
+    public function getListOfColumns() : array {
+        //TODO Edit Ares Core - DataObjectManagerFactory make Manager be able to access to SchemaBuilder to retrieve list of columns safely.
```
If you like to, you can create a PR for ares core here: https://github.com/Domexx/ares-habbo-backend

Stankman replied:
I will, I haven't edited nothing on Ares-Core so I will just leave it as a TODO.

shubrt commented on src/Permission/Service/CreateRankService.php:
```diff
@@ -0,0 +1,71 @@
+            );
+        }
+
+        if($this->permissionRepository->getDataObjectManager()->insert($data)) {            
```
Why don't use repository save method ?

Stankman replied:
Oh, okay that one because I'm inserting WHOLE rank structure. Since Arcturus doesn't have a proper Access/Commands permission structure, Ares has defined 'Permissions' (Ranks) as only: _id, name, prefix, prefix_color, etc..._ BUT 'Permission' (Rank) Model doesn't contain ALL acc_* | cmd_* columns, that's why I'm inserting all PermissionData. 

**Honestly, I haven't tested save() method inserting ALL columns I just thought that from logic.**

shubrt commented on src/Role/Entity/Contract/RoleInterface.php:
```diff
@@ -18,6 +18,7 @@ interface RoleInterface
     public const COLUMN_NAME = 'name';
     public const COLUMN_DESCRIPTION = 'description';
     public const COLUMN_STATUS = 'status';
+    public const COLUMN_IS_ROOT = 'isRoot';
```
Entity column name should be in snake case if possible (is_root).

shubrt commented on src/Role/Entity/RoleHierarchy.php:
```diff
@@ -20,6 +20,8 @@ class RoleHierarchy extends DataObject implements RoleHierarchyInterface
     /** @var string */
     public const TABLE = 'ares_roles_hierarchy';
 
+    // TODO ADD ORDER ID 
```
Is this already done ?

shubrt commented on src/Role/Repository/RoleRankRepository.php:
```diff
@@ -33,60 +35,49 @@ class RoleRankRepository extends BaseRepository {
+        ->where(RoleRankInterface::COLUMN_ROLE_ID, '=', $roleId)
+        ->orWhere(RoleRankInterface::COLUMN_RANK_ID, '=', $rankId);
+        //TODO I dont like this...
+        return $this->getOneBy($searchCriteria, true, false);
```
Do you want to change that todo ?

Stankman replied:
Honestly I don't remember why I didn't like that, maybe some Cache Issues that's why I don't like it I can just move that TODO onto DocBlock just as Dome said

shubrt commented on src/Role/Service/FetchUserPermissionService.php:
```diff
@@ -34,20 +36,17 @@ public function __construct(
      * @return CustomResponseInterface
      */
-    public function execute(int $rank): CustomResponseInterface
+    public function execute(int $rank_id): CustomResponseInterface
```
Please change the style of the variable to camel case.

shubrt commented on src/Role/Service/RoleTreeService.php:
```diff
@@ -0,0 +1,86 @@
+        $rootRole = $this->roleRepository->getRootRole();
+
+        if($rootRole == null) {
+            //throw error no root set.
```
Please create an exception for this case.

shubrt commented on src/Role/Service/UpdateChildRoleParentService.php:
```diff
@@ -0,0 +1,142 @@
+                        ->orderBy(RoleHierarchyInterface::COLUMN_ORDER_ID, 'DESC')
+                        ->first();
+
+        $lastOrderId = ($lastChild) ? (($lastChild->getChildRoleId() === $childRoleId) ? $lastChild->getOrderId() : $lastChild->getOrderId() + 1) : 1;
```
Please convert this short if statement into structured if statement for better readability

shubrt commented on src/User/Controller/AuthController.php:
```diff
@@ -107,12 +107,11 @@ public function register(Request $request, Response $response): Response
 
         $this->validationService->validate($parsedData, [
-            UserInterface::COLUMN_USERNAME => 'required|min:2|max:12|regex:/^[a-zA-Z\d]+$/',
+            UserInterface::COLUMN_USERNAME => 'required|min:3|max:12|regex:/^[a-zA-Z]+([\.:_\-=!\?]){0,5}([a-zA-Z0-9])+$/',
```
What's the reason for this additional regex check ?

Stankman replied:
Actually it would be cool to be able to add your own custom regular expression through configuration so that Owners/Developers can register Usernames as they want. In my case I wanted the possibility to add _.:_-=!_ characters in between letters.

shubrt commented on src/User/Controller/AuthController.php:
```diff
@@ -107,12 +107,11 @@ public function register(Request $request, Response $response): Response
             UserInterface::COLUMN_GENDER => 'required|default:M|regex:/[M.F]/',
-            UserInterface::COLUMN_PASSWORD => 'required|min:6',
-            'password_confirmation' => 'required|same:password'
+            UserInterface::COLUMN_PASSWORD => 'required|min:6','password_confirmation' => 'required|same:password'
```
What's the reason for this style change ?

Stankman replied:
For some reason, I changed many lines same way, I found it better to read and if lines are too large on VS Code I just go _Alt + Z_

shubrt commented on src/User/Service/Currency/UpdateCurrencyService.php:
```diff
@@ -41,27 +44,62 @@ public function __construct(
      */
     public function execute(int $userId, int $type, int $amount): void
```
Please try to split these function into smaller parts, to get some cleaner code and not that much clustered statements.

s1njar reviewed (changes requested):
Hi @RodolfoXDR ,
changes look really good thank you.
Please check these few comments.

Please also check the failed Codefactor checks below.

Thanks

Commit:
Turn 'is_root' to Snake Case

Commit:
Add VSCode Folder to Gitignore

Commit:
Changed 'if' structure

Commit:
Changed to Camel Case

Commit:
Removed TODO on RoleHierarchy

Commit:
RoleTreeService DONE + Optimizations

Commit:
UpdateCurrencyService Refactor

Commit:
Removed In-line TODO Comments

Commit:
Fixed Role Stuff and Added Change Rank by userID

Domexx commented on src/Role/Service/CreateChildRoleService.php:
```diff
@@ -62,10 +63,24 @@ public function execute(array $data): CustomResponseInterface
-        /** @var int $parentChildrenCount */
-        $parentChildrenCount = count($this->roleHierarchyRepository->getChildIds([$parentRoleId]));
+        /** @var RoleHierarchy $lastChild */
+        $lastChild = $this->roleHierarchyRepository->getDataObjectManager()
```
can you please outsource this into a custom function inside the given repository?

Domexx commented on src/Role/Service/UpdateRoleRankService.php:
```diff
@@ -49,19 +53,51 @@ public function execute(array $data): CustomResponseInterface
         /** @var int $roleId */
         $roleId = $data['role_id'];
+        
+        $searchCriteria = $this->roleRankRepository
```
same here, outsource in a custom function inside the repository

Commit:
Added $isCached Flags

Commit:
Shop Payments, RconCommands, Room and Guilds

Commit:
Changes on Description

Stankman commented:
- Added two permissions on initial SQL run: 

'override-ads'
'hide-leaderboard'

- [FIX] Added 'order_id' on table ares_roles_hierarchy on initial SQL run.

- [FIX] Changed value in ares_settings on 'register_enabled' from true to 1.

- Added two fetch on role-tree (staff page & control panel):
Staff Page Fetch doesn't contain hidden roles.
Control Panel Fetch contains hidden roles.

- Added Fetch function Role Permissions by Name

- Added Fetch function get Roles that have certain Permission Id

- Refactored RoleRank functions

- [FIX] 'Change Rank' always returned true.

- Refactored Top Leaderboards Hall of Fame

Commit:
Fixed other stuff and merge corrections

Commit:
Ares Refactored Routes and Rank Names
